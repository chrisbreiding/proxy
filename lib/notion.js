const rp = require('request-promise')

const getData = ({ notionPageId, notionToken }) => {
  return rp({
    uri: `https://api.notion.com/v1/blocks/${notionPageId}/children`,
    headers: {
      'Authorization': `Bearer ${notionToken}`,
      'Notion-Version': '2021-05-13',
    },
    json: true,
  })
}

module.exports = {
  getData,
}
