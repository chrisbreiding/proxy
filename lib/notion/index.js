const { addUpcomingWeek } = require('./upcoming-week')
const { getData } = require('./data')
const { onSocket } = require('./shopping')

module.exports = {
  addUpcomingWeek,
  getData,
  onSocket,
}
