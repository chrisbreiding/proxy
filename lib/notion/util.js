const rp = require('request-promise')

const makeRequest = ({ notionToken, body, method = 'get', path }) => {
  return rp({
    method,
    body,
    uri: `https://api.notion.com/v1/${path}`,
    headers: {
      'Authorization': `Bearer ${notionToken}`,
      'Notion-Version': '2021-05-13',
    },
    json: true,
  })
}

const getBlockChildren = ({ notionToken, pageId }) => {
  return makeRequest({ notionToken, path: `blocks/${pageId}/children` })
}

const appendBlockChildren = ({ notionToken, pageId, blocks }) => {
  return makeRequest({
    notionToken,
    method: 'patch',
    path: `blocks/${pageId}/children`,
    body: {
      children: blocks,
    },
  })
}

const updateBlock = ({ notionToken, blockId, block }) => {
  return makeRequest({
    notionToken,
    method: 'patch',
    path: `blocks/${blockId}`,
    body: block,
  })
}

const getPlainText = (withText) => {
  if (!withText || !withText.text) return

  return withText.text
  .map(({ plain_text }) => plain_text)
  .join('')
  .trim()
}

const weatherIconMap = {
  'clear-day': '☀️',
  'clear-night': '☀️',
  'rain': '☔️',
  'snow': '❄️',
  'sleet': '🌨',
  'wind': '💨',
  'fog': '🌫',
  'cloudy': '☁️',
  'partly-cloudy-day': '⛅️',
  'partly-cloudy-night': '⛅️',
  'default': '🌑',
}

const getWeatherIcon = (iconName) => {
  return weatherIconMap[iconName] || weatherIconMap.default
}

const makeTextPart = (content, color) => {
  const textPart = { text: { content } }

  if (color) {
    textPart.annotations = { color }
  }

  return textPart
}

module.exports = {
  getBlockChildren,
  appendBlockChildren,
  updateBlock,
  getPlainText,
  getWeatherIcon,
  makeTextPart,
}
