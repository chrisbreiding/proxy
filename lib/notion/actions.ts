import type express from 'express'
import { addUpcomingWeek } from './upcoming-week'
import { clearCompleted, deleteRecentlyCleared } from './clear-completed'
import { sendHtml, sendHtmlError } from './util/general'
import { promoteDay } from './promote-day'

function sendSuccess (res: express.Response, message: string) {
  sendHtml(res, 200,
    `<!DOCTYPE html>
    <html>
      <body>
        <h2 style="margin: 20px;">${message}<h2>
      </body>
    </html>`,
  )
}

function sendError (res: express.Response, error: any, message: string, statusCode = 500) {
  sendHtmlError({
    error,
    message,
    res,
    statusCode,
  })
}

export function action (req: express.Request, res: express.Response) {
  const action = req.query.action

  if (!action) {
    sendHtml(res, 400, '<p>A value for <em>action</em> must be provided in the query string</p>')

    return
  }

  const onSuccess = sendSuccess.bind(null, res)
  const onError = sendError.bind(null, res)

  switch (action) {
    case 'addUpcomingWeek':
      return addUpcomingWeek(req, onSuccess, onError)
    case 'clearCompleted':
      return clearCompleted(req, onSuccess, onError)
    case 'deleteRecentlyCleared':
      return deleteRecentlyCleared(req, onSuccess, onError)
    case 'promoteDay':
      return promoteDay(req, onSuccess, onError)
    default:
      sendHtml(res, 400, `<p>Action not supported: ${action}</p>`)
  }
}

