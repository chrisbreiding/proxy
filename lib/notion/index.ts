import { addUpcomingWeek, upcomingWeekView } from './upcoming-week'
import { getBlockChildren } from './util'
import { onSocket } from './shopping'

interface GetDataOptions {
  notionToken: string
  notionPageId: string
}

export function getNotionData ({ notionToken, notionPageId }: GetDataOptions) {
  return getBlockChildren({ notionToken, pageId: notionPageId })
}

export {
  addUpcomingWeek,
  upcomingWeekView,
  onSocket,
}
