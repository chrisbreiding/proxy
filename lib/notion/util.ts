import type { Method } from 'axios'
import dayjs from 'dayjs'
import type {
  ApiColor,
  BlockObjectResponse,
  BookmarkBlockObjectResponse,
  BulletedListItemBlockObjectResponse,
  CalloutBlockObjectResponse,
  ChildDatabaseBlockObjectResponse,
  ChildPageBlockObjectResponse,
  ColumnBlockObjectResponse,
  ColumnListBlockObjectResponse,
  DividerBlockObjectResponse,
  EmbedBlockObjectResponse,
  EquationBlockObjectResponse,
  FileBlockObjectResponse,
  Heading1BlockObjectResponse,
  Heading2BlockObjectResponse,
  Heading3BlockObjectResponse,
  ImageBlockObjectResponse,
  LinkPreviewBlockObjectResponse,
  LinkToPageBlockObjectResponse,
  ListBlockChildrenResponse,
  ListDatabasesResponse,
  NumberedListItemBlockObjectResponse,
  ParagraphBlockObjectResponse,
  PdfBlockObjectResponse,
  QuoteBlockObjectResponse,
  RichTextItemResponse,
  SyncedBlockBlockObjectResponse,
  TableBlockObjectResponse,
  TableOfContentsBlockObjectResponse,
  TableRowBlockObjectResponse,
  TemplateBlockObjectResponse,
  ToDoBlockObjectResponse,
  ToggleBlockObjectResponse,
  UnsupportedBlockObjectResponse,
  UpdateBlockBodyParameters,
  VideoBlockObjectResponse,
} from '@notionhq/client/build/src/api-endpoints'

import { clone, compact } from '../util/collections'
import { debug } from '../util/debug'
import { request } from '../util/network'

export const dateRegex = /(?:Sun|Mon|Tue|Wed|Thu|Fri|Sat), (\d{1,2}\/\d{1,2})/

export function getDateFromText (dateText: string) {
  const currentDate = dayjs()
  // originally assume the date's year matches the current year
  const date = dayjs(`${dateText}/${currentDate.year()}`, 'M/D/YYYY')
  // if the current month is after the date's month, we've crossed over years
  // and the date's year should be the next year. usually, it's because it's
  // currently December, but the date's month is January. this can be
  // generically applied, so it will work if it's currently June, but we're
  // trying to get the weather for April. we don't want past weather, only
  // future weather, so assume it's for the following year.
  if (currentDate.month() > date.month()) {
    return date.add(1, 'year')
  }

  return date
}

interface MakeRequestOptions {
  body?: object
  method?: Method
  notionToken: string
  path: string
}

function makeRequest<T> (options: MakeRequestOptions): Promise<T> {
  const { notionToken, body, path, method = 'get' } = options

  return request({
    method,
    body,
    url: `https://api.notion.com/v1/${path}`,
    headers: {
      'Authorization': `Bearer ${notionToken}`,
      'Notion-Version': '2022-06-28',
    },
  })
}

interface GetBlockChildrenOptions {
  notionToken: string
  pageId: string
}

export async function getBlockChildren ({ notionToken, pageId }: GetBlockChildrenOptions) {
  const { results } = await makeRequest<ListBlockChildrenResponse>({ notionToken, path: `blocks/${pageId}/children` }) as { results: BlockObjectResponse[] }

  return results.map(convertNotionBlockToOwnBlock)
}

interface GetBlockChildrenDeepOptions {
  filter?: (block: NotionBlock) => boolean
  includeId?: boolean
  notionToken: string
  pageId: string
}

export async function getBlockChildrenDeep ({ notionToken, pageId, filter, includeId }: GetBlockChildrenDeepOptions): Promise<NotionBlock[]> {
  const blocks = await getBlockChildren({ notionToken, pageId })
  const filteredBlocks = filter ? blocks.filter(filter) : blocks
  const blocksWithChildren = await Promise.all(filteredBlocks.map(async (block) => {
    if (block.has_children && block.type !== 'child_page') {
      block.children = await getBlockChildrenDeep({ notionToken, pageId: block.id, includeId })
    }

    return block
  }))

  return compact(blocksWithChildren)
}

type Content = BookmarkBlockObjectResponse['bookmark']
| BulletedListItemBlockObjectResponse['bulleted_list_item']
| CalloutBlockObjectResponse['callout']
| ChildDatabaseBlockObjectResponse['child_database']
| ChildPageBlockObjectResponse['child_page']
| ColumnBlockObjectResponse['column']
| ColumnListBlockObjectResponse['column_list']
| DividerBlockObjectResponse['divider']
| EmbedBlockObjectResponse['embed']
| EquationBlockObjectResponse['equation']
| FileBlockObjectResponse['file']
| Heading1BlockObjectResponse['heading_1']
| Heading2BlockObjectResponse['heading_2']
| Heading3BlockObjectResponse['heading_3']
| ImageBlockObjectResponse['image']
| LinkPreviewBlockObjectResponse['link_preview']
| LinkToPageBlockObjectResponse['link_to_page']
| NumberedListItemBlockObjectResponse['numbered_list_item']
| ParagraphBlockObjectResponse['paragraph']
| PdfBlockObjectResponse['pdf']
| QuoteBlockObjectResponse['quote']
| SyncedBlockBlockObjectResponse['synced_block']
| TableBlockObjectResponse['table']
| TableOfContentsBlockObjectResponse['table_of_contents']
| TableRowBlockObjectResponse['table_row']
| TemplateBlockObjectResponse['template']
| ToDoBlockObjectResponse['to_do']
| ToggleBlockObjectResponse['toggle']
| UnsupportedBlockObjectResponse['unsupported']
| VideoBlockObjectResponse['video']

export interface OwnBlock {
  children?: OwnBlock[] | NotionBlock[]
  type: BlockObjectResponse['type']
  content: Content
}

export interface NotionBlock extends OwnBlock {
  children?: NotionBlock[]
  has_children: boolean
  id: string
}

export type Block = OwnBlock | NotionBlock

function getRichText (block: OwnBlock): RichTextItemResponse[] | undefined {
  return 'rich_text' in block.content ? block.content.rich_text : undefined
}

interface MakeAppendRequestOptions {
  blocks: OutgoingBlock[]
  notionToken: string
  pageId: string
}

export function makeAppendRequest ({ notionToken, pageId, blocks }: MakeAppendRequestOptions) {
  return makeRequest<ListBlockChildrenResponse>({
    notionToken,
    method: 'patch',
    path: `blocks/${pageId}/children`,
    body: {
      children: blocks,
    },
  })
  // don't understand why this fails converage
  /* c8 ignore next */
}

interface AppendBlockChildrenOptions {
  blocks: OwnBlock[]
  notionToken: string
  pageId: string
}

export async function appendBlockChildrenDeep ({ notionToken, pageId, blocks }: AppendBlockChildrenOptions) {
  function moveChildren (blocks: NotionBlock[] | OwnBlock[]) {
    return blocks.map((block) => {
      const convertedBlock = convertBlockToOutgoingBlock(block)

      if (!block.children) return convertedBlock

      // @ts-ignore
      convertedBlock[block.type].children = moveChildren(block.children)

      return convertedBlock
    })
  }

  return makeAppendRequest({ notionToken, pageId, blocks: moveChildren(blocks) })
}

export async function appendBlockChildren ({ notionToken, pageId, blocks }: AppendBlockChildrenOptions) {
  let toAppend: OutgoingBlock[] = []

  for (const block of blocks) {
    const convertedBlock = convertBlockToOutgoingBlock(block)

    if (block.children) {
      const children = block.children

      toAppend.push(convertedBlock)

      const { results } = await makeAppendRequest({ notionToken, pageId, blocks: toAppend })
      toAppend = []

      await appendBlockChildren({
        notionToken,
        pageId: results[results.length - 1].id,
        blocks: children,
      })
    } else {
      toAppend.push(convertedBlock)
    }
  }

  if (toAppend.length) {
    return makeAppendRequest({
      notionToken,
      pageId,
      blocks: toAppend,
    })
  }
}

interface UpdateBlockOptions {
  notionToken: string
  blockId: string
  block: Block
}

export function updateBlock ({ notionToken, blockId, block }: UpdateBlockOptions) {
  const convertedBlock = convertBlockToOutgoingBlock(block)

  return makeRequest<void>({
    notionToken,
    method: 'patch',
    path: `blocks/${blockId}`,
    body: convertedBlock,
  })
}

interface UpdatePageOptions {
  notionToken: string
  pageId: string
  properties: {
    [key: string]: {
      date: {
        start: string
      }
    }
  }
}

export function updatePage ({ notionToken, pageId, properties }: UpdatePageOptions) {
  return makeRequest<void>({
    notionToken,
    body: { properties },
    method: 'patch',
    path: `pages/${pageId}`,
  })
}

interface QueryDatabasesOptions {
  notionToken: string
  databaseId: string
}

export function queryDatabases ({ notionToken, databaseId }: QueryDatabasesOptions) {
  return makeRequest<ListDatabasesResponse>({
    notionToken,
    method: 'post',
    path: `databases/${databaseId}/query`,
  })
}

export function convertNotionBlockToOwnBlock (block: BlockObjectResponse): NotionBlock {
  const type = block.type

  return {
    // @ts-ignore
    content: clone(block[type] as Content),
    has_children: block.has_children,
    id: block.id,
    type,
  }
  // don't understand why this fails converage
  /* c8 ignore next */
}

type OutgoingBlock = UpdateBlockBodyParameters & {
  object: string
  type: BlockObjectResponse['type']
}

export function convertBlockToOutgoingBlock (block: Block): OutgoingBlock {
  return {
    object: 'block',
    type: block.type,
    [block.type]: block.content,
  }
  // don't understand why this fails converage
  /* c8 ignore next */
}

interface MakeBlockOptions {
  annotations?: Partial<RichTextItemResponse['annotations']>
  children?: OwnBlock[] | NotionBlock[]
  text: string
  type: BlockObjectResponse['type']
}

export function makeBlock ({ text, type = 'paragraph', annotations, children }: MakeBlockOptions) {
  return {
    type,
    children,
    content: {
      rich_text: [
        {
          type: 'text',
          text: {
            content: text,
          },
          annotations,
          plain_text: text,
        },
      ],
    },
  } as OwnBlock
}

export function richTextToPlainText (richText: RichTextItemResponse[]) {
  return richText
  .map(({ plain_text }) => plain_text)
  .join('')
  .trim()
}

export function getBlockPlainText (block: Block) {
  const richText = getRichText(block)

  if (!richText) return

  return richTextToPlainText(richText)
}

interface RichTextItem {
  text: {
    content: string
    link?: { url: string } | null
  }
  type?: 'text'
  annotations?: {
    bold?: boolean
    italic?: boolean
    strikethrough?: boolean
    underline?: boolean
    code?: boolean
    color?: ApiColor
  }
}

export function makeTextPart (content: string, color?: ApiColor) {
  const textPart = { text: { content } } as RichTextItem

  if (color) {
    textPart.annotations = { color }
  }

  return textPart
}

export function getMonthNameFromIndex (monthIndex: number, short = false) {
  return dayjs().month(monthIndex).format(short ? 'MMM' : 'MMMM')
}

export function getMonths ({ short }: { short?: boolean } = {}) {
  return Array.from(new Array(12)).map((_, i) => {
    return getMonthNameFromIndex(i, short)
  })
}

/* c8 ignore start */
const getDisplayPrefix = (block: NotionBlock) => {
  switch (block.type) {
    case 'paragraph': return ''
    case 'heading_1': return '# '
    case 'heading_2': return '## '
    case 'heading_3': return '### '
    case 'to_do': return `[${'checked' in block.content && block.content.checked ? '✓' : ' '}] `
    case 'bulleted_list_item': return '• '
    case 'numbered_list_item': return '1. '
    case 'toggle': return '> '
    case 'code': return '<> '
    case 'quote': return '| '
    case 'callout': return '" '
    default: return ''
  }
}

function getDisplayText (block: NotionBlock) {
  if (block.type === 'child_page') {
    return `[${'title' in block.content ? block.content.title : '<untitled>'}]`
  }

  if (block.type === 'divider') {
    return '-----'
  }

  if (getRichText(block)) {
    return `${getDisplayPrefix(block)}${getBlockPlainText(block)}`
  }

  return `(${block.type})`
}

// displays block for logging/debugging purposes
export function displayBlocks (blocks: NotionBlock[]) {
  blocks.forEach((block) => {
    debug(`[${block.id}] ${getDisplayText(block)}${block.has_children ? ' (parent)' : ''}`)
  })
}
/* c8 ignore stop */
