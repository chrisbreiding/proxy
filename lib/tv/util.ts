import type { Method } from 'axios'
import { debug, debugVerbose } from '../util/debug'
import { getEnv } from '../util/env'
import { request } from '../util/network'

const apikey = getEnv('THETVDB_API_KEY')
const pin = getEnv('THETVDB_PIN')
export const baseUrl = 'https://api4.thetvdb.com'

interface LoginResult {
  status: string
  data: {
    token: string
  }
}

async function authenticate () {
  debugVerbose('authenticate')

  try {
    const { data, status } = await request({
      method: 'post',
      url: `${baseUrl}/v4/login`,
      body: { apikey, pin },
    }) as LoginResult

    if (!data?.token) {
      throw new Error(`Could not authenticate with TheTVDB, status: ${status}`)
    }

    return data.token as string
  } catch (error: any) {
    debug('Authentication failed:', error?.stack || error)

    throw error
  }
}

interface MakeRequestOptions {
  method?: Method
  params?: { [key: string]: string }
  path: string
  token?: string
}

export async function makeRequest (options: MakeRequestOptions) {
  const { method, params, path } = options
  const token = options.token || await authenticate()
  const url = `${baseUrl}/v4/${path}`

  debugVerbose('request: %o', { method, url, params, token })

  // TODO: handle error
  const result = await request({
    headers: {
      'Authorization': `Bearer ${token}`,
    },
    method,
    params,
    url,
  })

  result.token = token

  return result
}
