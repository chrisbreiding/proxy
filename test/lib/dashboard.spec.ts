import mockFs from 'mock-fs'
import nock from 'nock'
import {
  afterAll,
  afterEach,
  beforeEach,
  describe,
  expect,
  it,
  vi,
} from 'vitest'

process.env.API_KEY = 'key'
const token = process.env.APPLE_WEATHER_TOKEN = 'token'

import { startServer } from '../../index'
import { nockGetBlockChildren, nockNotion } from './notion/util'
import { fixtureContents, handleServer, weatherUrlBasePath } from '../util'

describe('lib/dashboard', () => {
  describe('GET /dashboard/:key', () => {
    handleServer(startServer)

    beforeEach(() => {
      vi.useFakeTimers()
      vi.setSystemTime(new Date(2022, 11, 28))
    })

    afterEach(() => {
      vi.useRealTimers()
    })

    afterAll(() => {
      mockFs.restore()
    })

    it('returns garage, quests, and beans data', async (ctx) => {
      nockGetBlockChildren('quests-id', { reply: { results: [] } })

      nock('https://weatherkit.apple.com')
      .matchHeader('authorization', `Bearer ${token}`)
      .get(`${weatherUrlBasePath}&dataSets=currentWeather,forecastHourly&hourlyStart=2022-12-28T05:00:00.000Z&hourlyEnd=2022-12-28T06:00:00.000Z`)
      .reply(200, fixtureContents('weather/weather'))

      nockNotion({
        method: 'post',
        path: '/v1/databases/beans-id/query',
        fixture: 'coffee/beans',
      })

      mockFs({
        'data': {
          'garage-data.json': JSON.stringify({ garage: 'data' }),
        },
      })

      const res = await ctx.request.get('/dashboard/key?location=lat,lng&notionToken=notion-token&notionQuestsId=quests-id&notionBeansId=beans-id')

      expect(res.status).to.equal(200)
      expect(res.body.garage).to.deep.equal({ data: { garage: 'data' } })
      expect(res.body.quests).to.deep.equal({ data: [] })
      expect(res.body.weather).toMatchInlineSnapshot(`
        {
          "data": {
            "apparentTemperature": 39.092,
            "icon": "clear",
            "precipAccumulation": 0.421259842519685,
            "precipProbability": 0.8,
            "summary": "Clear",
            "temperature": 23,
          },
        }
      `)
      expect(res.body.beans).toMatchSnapshot()
    })

    it('returns individual error instead of data', async (ctx) => {
      nockGetBlockChildren('quests-id', { reply: { results: [] } })

      nockNotion({
        method: 'post',
        path: '/v1/databases/beans-id/query',
        reply: { results: [] },
      })

      nock('https://weatherkit.apple.com')
      .matchHeader('authorization', `Bearer ${token}`)
      .get(`${weatherUrlBasePath}&dataSets=currentWeather,forecastHourly&hourlyStart=2022-12-28T05:00:00.000Z&hourlyEnd=2022-12-28T06:00:00.000Z`)
      .reply(500, { message: 'could not get weather' })

      mockFs({
        'data': {
          'garage-data.json': JSON.stringify({ garage: 'data' }),
        },
      })

      const res = await ctx.request.get('/dashboard/key?location=lat,lng&notionToken=notion-token&notionQuestsId=quests-id&notionBeansId=beans-id')

      expect(res.status).to.equal(200)
      expect(res.body.garage).to.deep.equal({ data: { garage: 'data' } })
      expect(res.body.quests).to.deep.equal({ data: [] })
      expect(res.body.beans).to.deep.equal({ data: [] })

      expect(res.body.weather.error).to.deep.equal({
        code: 'ERR_BAD_RESPONSE',
        message: 'Request failed with status code 500',
        name: 'AxiosError',
        response: { message: 'could not get weather' },
        status: 500,
        statusText: null,
      })
    })

    it('sends error if no location specified', async (ctx) => {
      const res = await ctx.request.get('/dashboard/key?notionToken=notion-token&notionQuestsId=quests-id&notionBeansId=beans-id')

      expect(res.status).to.equal(200)
      expect(res.body).to.deep.equal({ error: { message: 'Must include location in query' } })
    })

    it('sends error if no notionToken specified', async (ctx) => {
      const res = await ctx.request.get('/dashboard/key?location=lat,lng&notionQuestsId=quests-id&notionBeansId=beans-id')

      expect(res.status).to.equal(200)
      expect(res.body).to.deep.equal({ error: { message: 'Must include notionToken in query' } })
    })

    it('sends error if no notionQuestsId specified', async (ctx) => {
      const res = await ctx.request.get('/dashboard/key?location=lat,lng&notionToken=notion-token&notionBeansId=beans-id')

      expect(res.status).to.equal(200)
      expect(res.body).to.deep.equal({ error: { message: 'Must include notionQuestsId in query' } })
    })

    it('sends error if no notionBeansId specified', async (ctx) => {
      const res = await ctx.request.get('/dashboard/key?location=lat,lng&notionToken=notion-token&notionQuestsId=quests-id')

      expect(res.status).to.equal(200)
      expect(res.body).to.deep.equal({ error: { message: 'Must include notionBeansId in query' } })
    })

    it('sends 403 if key does not match', async (ctx) => {
      const res = await ctx.request.get('/dashboard/nope')

      expect(res.status).to.equal(403)
    })
  })
})
