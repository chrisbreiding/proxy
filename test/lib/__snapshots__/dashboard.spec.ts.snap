// Vitest Snapshot v1

exports[`lib/dashboard > GET /dashboard/:key > returns garage, quests, and beans data 2`] = `
{
  "data": [
    {
      "archived": false,
      "cover": null,
      "created_by": {
        "id": "45d332f7-e2ab-4213-b92a-ce937311866b",
        "object": "user",
      },
      "created_time": "2023-02-02T15:30:00.000Z",
      "icon": null,
      "id": "a68c021f-2404-48fe-8038-68fab09bbf82",
      "last_edited_by": {
        "id": "45d332f7-e2ab-4213-b92a-ce937311866b",
        "object": "user",
      },
      "last_edited_time": "2023-02-02T15:46:00.000Z",
      "object": "page",
      "parent": {
        "database_id": "90877700-6150-4875-8fd5-fa60ef8f2912",
        "type": "database_id",
      },
      "properties": {
        "Active": {
          "checkbox": true,
          "id": "iidH",
          "type": "checkbox",
        },
        "Bag Description": {
          "id": "%3DmOo",
          "rich_text": [
            {
              "annotations": {
                "bold": false,
                "code": false,
                "color": "default",
                "italic": false,
                "strikethrough": false,
                "underline": false,
              },
              "href": null,
              "plain_text": "Full-bodied, Chocolatey, Roasty",
              "text": {
                "content": "Full-bodied, Chocolatey, Roasty",
                "link": null,
              },
              "type": "text",
            },
          ],
          "type": "rich_text",
        },
        "Brand": {
          "id": "%3BQLD",
          "select": {
            "color": "green",
            "id": "015a9183-6281-48fc-93e6-78ff57c10304",
            "name": "La Columbe",
          },
          "type": "select",
        },
        "Decaffeination": {
          "id": "KvVG",
          "select": null,
          "type": "select",
        },
        "Get Again?": {
          "id": "uZEf",
          "select": null,
          "type": "select",
        },
        "Grind": {
          "id": "iG%7DL",
          "rich_text": [
            {
              "annotations": {
                "bold": false,
                "code": false,
                "color": "default",
                "italic": false,
                "strikethrough": false,
                "underline": false,
              },
              "href": null,
              "plain_text": "3",
              "text": {
                "content": "3",
                "link": null,
              },
              "type": "text",
            },
          ],
          "type": "rich_text",
        },
        "Last Drank": {
          "date": null,
          "id": "%5Cr%3Cr",
          "type": "date",
        },
        "Link": {
          "id": "%5DNVN",
          "type": "url",
          "url": "https://example.com/regular-coffee",
        },
        "Name": {
          "id": "title",
          "title": [
            {
              "annotations": {
                "bold": false,
                "code": false,
                "color": "default",
                "italic": false,
                "strikethrough": false,
                "underline": false,
              },
              "href": null,
              "plain_text": "Rise and Shine",
              "text": {
                "content": "Rise and Shine",
                "link": null,
              },
              "type": "text",
            },
          ],
          "type": "title",
        },
        "Notes": {
          "id": "~e%3Fh",
          "rich_text": [],
          "type": "rich_text",
        },
        "Origin": {
          "id": "DO%7BI",
          "multi_select": [
            {
              "color": "orange",
              "id": "b97d1567-7f7b-40bf-b108-c5f5ebb72a8a",
              "name": "Brazil",
            },
            {
              "color": "default",
              "id": "01a7498d-a39b-44ef-8d3a-4ba009561bd8",
              "name": "Colombia",
            },
            {
              "color": "pink",
              "id": "5cfb5d2f-3cfb-4248-9746-6b0de9872e00",
              "name": "Honduras",
            },
            {
              "color": "purple",
              "id": "a1a3bb6e-b154-412a-b3ee-917e8c06e0ee",
              "name": "Mexico",
            },
          ],
          "type": "multi_select",
        },
        "Roast": {
          "id": "WIW%3A",
          "select": {
            "color": "brown",
            "id": "b522a619-af6d-4b1c-9c51-0660977429a8",
            "name": "Dark",
          },
          "type": "select",
        },
        "Strength": {
          "id": "h%3BRj",
          "select": {
            "color": "blue",
            "id": "p=vB",
            "name": "Regular",
          },
          "type": "select",
        },
        "Temperature": {
          "id": "%3DmO0",
          "rich_text": [
            {
              "annotations": {
                "bold": false,
                "code": false,
                "color": "default",
                "italic": false,
                "strikethrough": false,
                "underline": false,
              },
              "href": null,
              "plain_text": "200",
              "text": {
                "content": "200",
                "link": null,
              },
              "type": "text",
            },
          ],
          "type": "rich_text",
        },
        "Type": {
          "id": "OYoC",
          "multi_select": [],
          "type": "multi_select",
        },
      },
      "url": "https://www.notion.so/regular",
    },
    {
      "archived": false,
      "cover": null,
      "created_by": {
        "id": "45d332f7-e2ab-4213-b92a-ce937311866b",
        "object": "user",
      },
      "created_time": "2023-01-31T01:42:00.000Z",
      "icon": null,
      "id": "ec763e6f-5936-4ee0-a1b7-b148cf1f5288",
      "last_edited_by": {
        "id": "45d332f7-e2ab-4213-b92a-ce937311866b",
        "object": "user",
      },
      "last_edited_time": "2023-02-02T21:45:00.000Z",
      "object": "page",
      "parent": {
        "database_id": "90877700-6150-4875-8fd5-fa60ef8f2912",
        "type": "database_id",
      },
      "properties": {
        "Active": {
          "checkbox": true,
          "id": "iidH",
          "type": "checkbox",
        },
        "Bag Description": {
          "id": "%3DmOo",
          "rich_text": [],
          "type": "rich_text",
        },
        "Brand": {
          "id": "%3BQLD",
          "select": {
            "color": "yellow",
            "id": "0e9f4817-922b-41ec-9935-bb5c3a5f2271",
            "name": "Sprouts",
          },
          "type": "select",
        },
        "Decaffeination": {
          "id": "KvVG",
          "select": null,
          "type": "select",
        },
        "Get Again?": {
          "id": "uZEf",
          "select": {
            "color": "green",
            "id": "wCFK",
            "name": "Yes",
          },
          "type": "select",
        },
        "Grind": {
          "id": "iG%7DL",
          "rich_text": [
            {
              "annotations": {
                "bold": false,
                "code": false,
                "color": "default",
                "italic": false,
                "strikethrough": false,
                "underline": false,
              },
              "href": null,
              "plain_text": "42",
              "text": {
                "content": "42",
                "link": null,
              },
              "type": "text",
            },
          ],
          "type": "rich_text",
        },
        "Last Drank": {
          "date": null,
          "id": "%5Cr%3Cr",
          "type": "date",
        },
        "Link": {
          "id": "%5DNVN",
          "type": "url",
          "url": "https://example.com/decaf-coffee",
        },
        "Name": {
          "id": "title",
          "title": [
            {
              "annotations": {
                "bold": false,
                "code": false,
                "color": "default",
                "italic": false,
                "strikethrough": false,
                "underline": false,
              },
              "href": null,
              "plain_text": "Sleepy Time",
              "text": {
                "content": "Sleepy Time",
                "link": null,
              },
              "type": "text",
            },
          ],
          "type": "title",
        },
        "Notes": {
          "id": "~e%3Fh",
          "rich_text": [
            {
              "annotations": {
                "bold": false,
                "code": false,
                "color": "default",
                "italic": false,
                "strikethrough": false,
                "underline": false,
              },
              "href": null,
              "plain_text": "Yummy yummy",
              "text": {
                "content": "Yummy yummy",
                "link": null,
              },
              "type": "text",
            },
          ],
          "type": "rich_text",
        },
        "Origin": {
          "id": "DO%7BI",
          "multi_select": [],
          "type": "multi_select",
        },
        "Roast": {
          "id": "WIW%3A",
          "select": {
            "color": "brown",
            "id": "b522a619-af6d-4b1c-9c51-0660977429a8",
            "name": "Dark",
          },
          "type": "select",
        },
        "Strength": {
          "id": "h%3BRj",
          "select": {
            "color": "yellow",
            "id": ":<lP",
            "name": "Decaf",
          },
          "type": "select",
        },
        "Temperature": {
          "id": "%3DmO0",
          "rich_text": [
            {
              "annotations": {
                "bold": false,
                "code": false,
                "color": "default",
                "italic": false,
                "strikethrough": false,
                "underline": false,
              },
              "href": null,
              "plain_text": "195",
              "text": {
                "content": "195",
                "link": null,
              },
              "type": "text",
            },
          ],
          "type": "rich_text",
        },
        "Type": {
          "id": "OYoC",
          "multi_select": [],
          "type": "multi_select",
        },
      },
      "url": "https://www.notion.so/decaf",
    },
  ],
}
`;
